Generate Stack Traces for bad http responses

dims opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

When we spit out a http reponse with a `0` http status code we
should log a trace back so we can easily find where things went wrong.

Change-Id: Ibb604d724fc8b660a2f16c7d14e5c0ad4932e4d0

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

dims commented:
/assign @liggitt

dims commented:
/test pull-kubernetes-integration

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:
```diff
@@ -33,14 +33,27 @@ func WithPanicRecovery(handler http.Handler) http.Handler {
 	})
 }
 
+func BadHttpResponse() httplog.StacktracePred {
```
avoid exporting this if you can

liggitt commented:
given that there were no panics logged when we saw this response, the stack trace here is not likely to give us any stack info down into the nested handler that was actually responsible for returning early without writing a status code or content type

dims commented:
the `Log()` in the `defer httplog.NewLogged(req, &w).StacktraceWhen(badHttpResponse()).Log()` is called when the actual serialization to stream takes place. so if at that point the http status is `0` or the `content-type` is not set then we will see a trace back.

worst case, we will not hit anything and we can revert this diagnostic change back...

liggitt commented:
let's see if we can get any info in this PR before merging... would prefer not to change the signature if possible

dims commented:
/test pull-kubernetes-integration

dims commented:
/test all

dims commented:
/test all

dims commented:
@sttts @liggitt given the low frequency of hits ... we won't catch the problem using `/test all` here. Is there something else we can do?

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=0-length%20response

dims commented:
/test all

wojtek-t commented:
> given that there were no panics logged when we saw this response, the stack trace here is not likely to give us any stack info down into the nested handler that was actually responsible for returning early without writing a status code or content type

+100 - my thoughts exactly
I'm not sure this will provide us more details.

dims commented:
@wojtek-t @liggitt fine. i'll just continue to investigate with stuff here

/hold

dims commented:
/test all

fedebongio commented:
/cc @logicalhan

dims committed:
Generate Stack Traces for http response with status code zero

When we spit out a http reponse with a `0` http status code we
should log a trace back so we can easily find where things went wrong.

Change-Id: Ic2aadec3a3de85fbdf64da66d6d12e3c631f409d

dims commented:
/assign @wojtek-t

dims commented:
/hold cancel

liggitt commented:
/lgtm
/approve

logicalhan reviewed (approved):
/lgtm

wojtek-t commented:
LGTM - thanks @dims 
Let's hope we will get some results from it soon.

liggitt commented:
no... https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/70439/pull-kubernetes-integration/40019/ failed:

```
--- FAIL: Test202StatusCode (3.75s)
    apiserver_test.go:140: Failed to create rs: 0-length response with status code: 200 and content type: 
```

and the "Bad HTTP Response" message did not get logged
